lib/src/execution/execution.cc

```diff
@@ -121,157 +127,224 @@ namespace facter { namespace execution {
-                output.append(buffer);
+            // Skip empty lines
+            if (buffer.empty()) {
                 continue;
             }
 
-            // Find the last newline, because anything after may not be a complete line.
-            auto lastNL = buffer.find_last_of("\n\r");
-            if (lastNL == string::npos) {
-                // No newline found, so keep appending and continue.
-                output.append(buffer);
-                continue;
+            // Log the line to the output logger
+            if (LOG_IS_DEBUG_ENABLED()) {
+                log(logger, log_level::debug, buffer);
             }
 
-            // Make a range for iterating through lines.
-            auto str_range = make_pair(buffer.begin(), buffer.begin()+lastNL);
-            auto line_iterator = boost::make_iterator_range(
-                make_split_iterator(str_range, token_finder(is_any_of("\n\r"), token_compress_on)),
-                split_iterator<string::iterator>());
+            // Pass the line to the callback
+            bool finished = !callback(buffer);
 
-            for (auto &line : line_iterator) {
-                // The previous trailing data is picked up by default.
-                output.append(line.begin(), line.end());
+            // Clear the line for the next iteration
+            buffer.clear();
 
-                if (trim_output) {
-                    boost::trim(output);
-                }
+            // Break out if finished processing
+            if (finished) {
+                return false;
+            }
+        }
 
-                // Skip empty lines
-                if (output.empty()) {
-                    continue;
-                }
+        // Save the new trailing data
+        buffer.assign(data.begin()+lastNL, data.end());
+        return true;
+    }
 
-                // Log the line to the output logger
-                if (LOG_IS_DEBUG_ENABLED()) {
-                    log(logger, log_level::debug, output);
-                }
+    tuple<string, string> process_streams(bool trim, function<bool(string&)> const& stdout_callback, function<bool(string&)> const& stderr_callback, function<void(function<bool(string const&)>, function<bool(string const&)>)> const& read_streams)
+    {
+        // Get a special logger used specifically for child process output
+        static const string stdout_logger = "|";
+        static const string stderr_logger = "!!!";
 
-                // Pass the line to the callback
-                if (!callback(output)) {
+        // Buffers for all of the output or partial line output
+        string stdout_buffer;
+        string stderr_buffer;
+
+        // Read the streams
+        read_streams(
+            [&](string const& data) {
+                if (!process_data(trim, data, stdout_buffer, stdout_logger, stdout_callback)) {
                     LOG_DEBUG("completed processing output; closing child pipe.");
```

Comment:
It should really be "closing child pipes", as both pipes will close when we stop processing. Will fix.